File: backend/ingest.py
import os
import asyncio
from dotenv import load_dotenv
from supabase import create_client, Client
import google.generativeai as genai
from pypdf import PdfReader
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Load environment variables
load_dotenv()

# Configuration
SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_KEY")
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")

# Initialize clients
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)
genai.configure(api_key=GEMINI_API_KEY)

PDF_PATH = "../Fundamentals of CNC Machining.pdf"

async def setup_database():
    """
    Sets up the database with the vector extension and documents table.
    """
    print("Setting up database...")
    
    # Enable vector extension
    try:
        # We can't easily run arbitrary SQL via the JS/Python client without a stored procedure 
        # or using the SQL editor in the dashboard. 
        # However, we can try to use the `rpc` call if a function exists, 
        # or just assume the user might need to run this in the dashboard if this fails.
        # For now, let's try to proceed. If the table doesn't exist, we might need the user 
        # to run a SQL snippet in their Supabase dashboard.
        pass 
    except Exception as e:
        print(f"Note: Ensure 'vector' extension is enabled in Supabase. Error: {e}")

    # Ideally, we would create the table here. 
    # Since the standard client limits DDL, we will print instructions if insertion fails.
    pass

def get_embedding(text):
    """Generates embedding using Gemini."""
    result = genai.embed_content(
        model="models/text-embedding-004",
        content=text,
        task_type="retrieval_document",
        title="CNC Knowledge"
    )
    return result['embedding']

async def process_pdf():
    print(f"Reading {PDF_PATH}...")
    reader = PdfReader(PDF_PATH)
    text = ""
    for page in reader.pages:
        text += page.extract_text() + "\n"
    
    print(f"Total characters: {len(text)}")

    # Split text into chunks
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=1000,
        chunk_overlap=200,
        length_function=len,
    )
    chunks = text_splitter.split_text(text)
    print(f"Total chunks: {len(chunks)}")

    # Process and insert chunks
    print("Generating embeddings and storing in Supabase...")
    
    for i, chunk in enumerate(chunks):
        try:
            embedding = get_embedding(chunk)
            
            data = {
                "content": chunk,
                "metadata": {"source": "Fundamentals of CNC Machining.pdf", "chunk_index": i},
                "embedding": embedding
            }
            
            # Insert into 'documents' table
            response = supabase.table("documents").insert(data).execute()
            
            if i % 10 == 0:
                print(f"Processed {i}/{len(chunks)} chunks...")
                
        except Exception as e:
            print(f"Error processing chunk {i}: {e}")
            # If error is about table not found, we need to notify user
            if "relation \"documents\" does not exist" in str(e):
                print("\nCRITICAL: Table 'documents' does not exist.")
                return

    print("Ingestion complete!")

if __name__ == "__main__":
    asyncio.run(process_pdf())

File: backend/ingest.js
const { GoogleGenerativeAI } = require('@google/generative-ai');
const { RecursiveCharacterTextSplitter } = require('@langchain/textsplitters');
const { createClient } = require('@supabase/supabase-js');
const pdf = require('pdf-parse');
const fs = require('fs');
const path = require('path');
require('dotenv').config();

// Configuration
const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY;
const GEMINI_API_KEY = process.env.GEMINI_API_KEY;

// Initialize clients
const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
const genAI = new GoogleGenerativeAI(GEMINI_API_KEY);
const model = genAI.getGenerativeModel({ model: "text-embedding-004" });

// PDF PATH - இங்கே மாற்றவும்!
const PDF_PATH = path.join(__dirname, '../Fundamentals of CNC Machining.pdf');

async function getEmbedding(text, retries = 5) {
    for (let i = 0; i < retries; i++) {
        try {
            const result = await model.embedContent(text);
            return result.embedding.values;
        } catch (error) {
            if (error.status === 429 || error.message?.includes('429')) {
                const delay = Math.min(1000 * Math.pow(2, i), 20000);
                console.log(`Rate limit hit. Retrying in ${delay}ms...`);
                await new Promise(resolve => setTimeout(resolve, delay));
            } else {
                console.error("Error generating embedding:", error.message);
                return null;
            }
        }
    }
    console.error("Max retries reached for embedding.");
    return null;
}

async function processPdf() {
    console.log(`📄 Reading PDF: ${PDF_PATH}...`);

    if (!fs.existsSync(PDF_PATH)) {
        console.error("❌ Error: PDF file not found at", PDF_PATH);
        console.log("\n📝 How to use:");
        console.log("1. Copy your PDF file to the project folder");
        console.log("2. Update PDF_PATH in ingest.js (line 20)");
        console.log("3. Run: node ingest.js");
        return;
    }

    let data;
    try {
        const dataBuffer = fs.readFileSync(PDF_PATH);
        data = await pdf(dataBuffer);
        console.log(`✅ PDF loaded successfully!`);
        console.log(`📊 Total pages: ${data.numpages}`);
        console.log(`📝 Total characters: ${data.text.length}`);
    } catch (e) {
        console.error("❌ Error parsing PDF:", e);
        return;
    }

    const textContent = data.text;

    // Split into chunks (1000 characters for good balance)
    const splitter = new RecursiveCharacterTextSplitter({
        chunkSize: 1000,
        chunkOverlap: 100,
    });

    const chunks = await splitter.splitText(textContent);
    console.log(`📦 Split into ${chunks.length} chunks`);
    console.log('🔄 Generating embeddings and storing in Supabase...\n');

    let successCount = 0;
    let failCount = 0;

    for (let i = 0; i < chunks.length; i++) {
        const chunkContent = chunks[i];

        // Skip empty chunks
        if (!chunkContent.trim()) {
            console.log(`⏭️  Skipping empty chunk ${i + 1}`);
            continue;
        }

        const embedding = await getEmbedding(chunkContent);

        if (embedding) {
            const { error } = await supabase
                .from('documents')
                .insert({
                    content: chunkContent,
                    metadata: {
                        source: path.basename(PDF_PATH),
                        chunk_index: i,
                        total_chunks: chunks.length
                    },
                    embedding: embedding
                });

            if (error) {
                console.error(`❌ Error inserting chunk ${i + 1}:`, error.message);
                failCount++;
            } else {
                successCount++;
                if (successCount % 10 === 0) {
                    console.log(`✅ Processed ${successCount}/${chunks.length} chunks...`);
                }
            }
        } else {
            failCount++;
        }

        // Small delay to avoid rate limits
        await new Promise(resolve => setTimeout(resolve, 200));
    }

    console.log('\n📊 Ingestion Summary:');
    console.log(`✅ Success: ${successCount} chunks`);
    console.log(`❌ Failed: ${failCount} chunks`);
    console.log(`📈 Success rate: ${((successCount / chunks.length) * 100).toFixed(1)}%`);
    console.log('\n🎉 Ingestion complete!');
}

processPdf();
